subprojects/core/src/test/groovy/org/gradle/api/internal/project/taskfactory/AnnotationProcessingTaskFactoryTest.groovy

```diff
@@ -210,6 +210,49 @@ class AnnotationProcessingTaskFactoryTest extends AbstractProjectBuilderSpec {
         0 * _
     }
 
+    def "uses IncrementalTaskInputsMethod when it is overriden"() {
+        given:
+        def incrementalTaskInputsAction = Mock(Action)
+        def inputChangesAction = Mock(Action)
+        def task = expectTaskCreated(AnnotationProcessingTasks.TaskOverridingDeprecatedIncrementalChangesActions, incrementalTaskInputsAction, inputChangesAction)
+
+        when:
+        execute(task)
+
+        then:
+        1 * incrementalTaskInputsAction.execute(_ as IncrementalTaskInputs)
+        1 * inputChangesAction.execute(_ as InputChanges)
+        0 * _
+    }
+
+    def "uses InputChanges method when two actions are present on the same class"() {
+        given:
+        def incrementalTaskInputsAction = Mock(Action)
+        def inputChangesAction = Mock(Action)
+        def task = expectTaskCreated(AnnotationProcessingTasks.TaskWithOverloadedDeprecatedIncrementalAndInputChangesActions, incrementalTaskInputsAction, inputChangesAction)
+
+        when:
+        execute(task)
+
+        then:
+        1 * inputChangesAction.execute(_ as InputChanges)
+        0 * _
+    }
+
+    def "uses overridden InputChanges when two actions are present on the base class"() {
+        given:
+        def incrementalTaskInputsAction = Mock(Action)
+        def inputChangesAction = Mock(Action)
+        def task = expectTaskCreated(AnnotationProcessingTasks.TaskOverridingInputChangesActions, incrementalTaskInputsAction, inputChangesAction)
+
+        when:
+        execute(task)
+
+        then:
+        1 * inputChangesAction.execute(_ as InputChanges)
+        0 * _
+    }
```

Comment:
These tests could be simplified a bit. You could set a single `Action` and check the type that it gets called with (`1 * paramAction.execute(_ as InputChanges)` vs `1 * paramAction.execute(_ as IncrementalTaskInputs)`). Another solution would be to expose the type of the parameter handled in a property of the task, and observe that. The latter would require no mock to be passed to the task construction.

Reply:
Using one `Action` now.